fairscale/experimental/nn/sync_batchnorm.py

```diff
@@ -0,0 +1,170 @@
+
+import torch
+import torch.distributed as dist
+from torch.distributed import ProcessGroup
+
+if torch.__version__.split(".")[:2] >= ["1", "8"]:
+    from torch.distributed import all_reduce
+else:
+    # Copied from https://github.com/pytorch/pytorch/blob/v1.8.1/torch/distributed/nn/functional.py
+    class _AllReduce(torch.autograd.Function):
+        @staticmethod
+        def forward(ctx, op, group, tensor):  # type: ignore
+            ctx.group = group
+            ctx.op = op
+            tensor = tensor.clone()
+            dist.all_reduce(tensor, op=op, group=group)
+            return tensor
+
+        @staticmethod
+        def backward(ctx, grad_output):  # type: ignore
+            return (None, None) + (_AllReduce.apply(ctx.op, ctx.group, grad_output),)
+
+    def all_reduce(tensor, op=dist.ReduceOp.SUM, group=dist.group.WORLD):  # type: ignore
+        return _AllReduce.apply(op, group, tensor)
+
+
+def _forward(
+    input: torch.Tensor,
+    affine: bool,
+    track_running_stats: bool,
+    mean: torch.Tensor,
+    var: torch.Tensor,
+    momentum: float,
+    eps: float,
+    weight_param: torch.Tensor,
+    bias_param: torch.Tensor,
+    running_mean: torch.Tensor,
+    running_var: torch.Tensor,
+    total_count: torch.Tensor,
+) -> torch.Tensor:
+    scale = torch.rsqrt(var + eps)
+    bias = -mean * scale
+    if affine:
+        scale = scale * weight_param.reshape(mean.shape)
+        bias = bias + bias_param.reshape(mean.shape)
+
+    if track_running_stats:
+        with torch.no_grad():
+            unbiased_var = var * (total_count / (total_count - 1))
+            running_mean += momentum * (mean.reshape(-1) - running_mean)
+            running_var += momentum * (unbiased_var.reshape(-1) - running_var)
+
+    return input * scale + bias
+
+
+if torch.__version__.split(".")[:2] >= ["1", "7"]:
+    _forward = torch.jit.script(_forward)  # type: ignore
+
+
+class SyncBatchNorm(torch.nn.BatchNorm2d):
+    """
+    Fast re-implementation of ``torch.nn.SyncBatchNorm`` that can achieve a speedup
+    of 5x or more over the default implementation depending on size of the input
+    and number of distributed world.
+    """
+
+    def __init__(
+        self, *args: Tuple[Any, ...], process_group: Optional[ProcessGroup] = None, **kwargs: Dict[str, Any]
+    ) -> None:
+        super().__init__(*args, **kwargs)  # type: ignore
+        self._process_group = process_group if process_group is not None else dist.group.WORLD
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:  # type: ignore
+        if not dist.is_initialized() or not self.training:
+            return super().forward(input)
+
+        dim = [d for d in range(input.ndim) if d != 1]
+        count = torch.full((1,), input.numel() // input.size(1), device=input.device, dtype=input.dtype)
+        total_count = count.clone()
+        handle = all_reduce(total_count, group=self._process_group, async_op=True)
```

Comment:
I can see that this might be useful when we have a partial batch. Curious, are there other reasons why we need this allreduce?

Reply:
Yeah, the primary motivation is to handle partial batches. Another case would be when batch sizes are not the same. This can sometimes happen in NLP due to variable sequence length although we don't use BatchNorm in NLP.

I am contemplating adding a batch-size argument that would avoid this all_reduce and assert if their is a non-matching batch. But this all_reduce does not impact performance that much since it is just a single float and is async. Although, it could matter in a network with high-latency.

Reply:
I think doing the allreduce might be better than requiring an explicit batch size argument.